ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveFilterProjectTransposeRule.java

```diff
@@ -339,6 +339,11 @@ private void check(Filter filter) {
       final RexNode filterCondition = simplify.simplify(filter.getCondition());
 
       final Set<Integer> inputRefs = HiveCalciteUtil.getInputRefs(newCondition);
+      // if the new IS NOT NULL has no input ref, there is redundancy here, bail out
+      if (inputRefs.isEmpty()) {
```

Comment:
this is interesting...I think we shouldn't have the  `IS NOT NULL` at all.... why don't we run `RexSimplify` in `unknownAsFalse` mode?

we may make that change independently from this change - but `uAF` could probably identify more opportunities in case the outer is not an `IS NOT NULL`

Reply:
I am tackling the use of `RexSimplify` in this rule in this other PR: https://github.com/apache/hive/pull/2840
The majority of the test diff are (expected) changes in the plan, but there are also few real failures I need to investigate, but you can check if what is done there matches what you have in mind.

Reply:
This method is invoked only for `IS NOT NULL` calls, so we can't have other kind of calls here.

This said, I agree about trying to use more `RexSimplify` with `uAF` mode when we create/modify predicates in our CBO rules, https://github.com/apache/hive/pull/2840 is heading into that direction for`HiveJoinPushTransitivePredicatesRule`, for instance.

It's probably a good idea to open a separate Jira ticket for revising our normalization/simplification strategy in CBO rules.

Reply:
there are a couple of ways to fix things like this:
* we could call `RexSimplify` everywhere - but I don't really like that idea as it exposes unneccessary complexity - at places where a higher level objective is targeted
* use the `RelBuilder` which is usually calling the simplifier here and there
* use a rule which runs simplifiaction - it could be that these rules should be accompanied by `ReduceExpressionsRule` ?

Reply:
For the `RelBuilder` use I think [HiveFilterMergeRule](https://github.com/apache/hive/blob/cadf5f731c8cb0a4217d50de8a9b216eaaed1ea1/ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveFilterMergeRule.java#L54) is a good example, and it hides all the complexity.

However, when dealing directly with newly created `RexNode`s I am not sure how to achieve that consistently.

Regarding option 3, it seems that at least in the case reported here it was not enough: https://github.com/apache/hive/pull/2840. My impression is that, if we don't simplify before hand, in some cases, `RexSimplify.simplify` is not always able to perform the simplification afterwards, but I need to have a better look.